Palace/Book/UI/AudiobookBookmarkBusinessLogic.swift

```diff
@@ -0,0 +1,230 @@
+      for bookmark in unsyncedLocalBookmarks {
+        let data = bookmark.toData()
+        guard let locationString = String(data: data, encoding: .utf8) else {
+          continue
+        }
+        
+        if let annotationId = try await annotationsManager.postAudiobookBookmark(forBook: self.book.identifier, selectorValue: locationString) {
+          if let updatedBookmark = bookmark.copy() as? ChapterLocation {
+            updatedBookmark.annotationId = annotationId
+            replace(oldLocation: bookmark, with: updatedBookmark)
+          }
+        }
+      }
+      
+      fetchServerBookmarks { [weak self] remoteBookmarks in
+        guard let self = self, !remoteBookmarks.isEmpty else {
+          return
+        }
+        
+        var updatedLocalBookmarks = self.fetchLocalBookmarks()
+        
+        // 2. Download server bookmarks that are not stored locally
+        let unsyncedRemoteBookmarks = remoteBookmarks.filter { remoteBookmark in
+          !updatedLocalBookmarks.contains(where: { localBookmark in
+            localBookmark.isSimilar(to: remoteBookmark)
+          })
+        }
+        
+        unsyncedRemoteBookmarks.forEach {
+          if let genericLocation = $0.toTPPBookLocation() {
+            self.registry.addOrReplaceGenericBookmark(genericLocation, forIdentifier: self.book.identifier)
+          }
+        }
+        
+        updatedLocalBookmarks = self.fetchLocalBookmarks()
+        
+        // 3. Delete local bookmarks that have been deleted from the server
+        let localBookmarksToDelete = remoteBookmarks.filter { remoteBookmark in
+          updatedLocalBookmarks.contains(where: { localBookmark in
+            localBookmark.isSimilar(to: remoteBookmark) && localBookmark.annotationId.isEmpty
+          })
+        }
+        
+        localBookmarksToDelete.forEach { self.deleteBookmark(at: $0) { _ in } }
+        
+        // 4. Delete server bookmarks that have been deleted locally
+        let serverBookmarksToDelete = updatedLocalBookmarks.filter { localBookmark in
+          !remoteBookmarks.contains(where: { remoteBookmark in
+            remoteBookmark.isSimilar(to: localBookmark)
+          }) && !localBookmark.annotationId.isEmpty
+        }
+        
+        for bookmark in serverBookmarksToDelete {
+          if !bookmark.annotationId.isEmpty {
+            self.annotationsManager.deleteBookmark(annotationId: bookmark.annotationId) { _ in }
+          }
+        }
+
+        self.isSyncing = false
+        completion?()
+      }
+    }
+  }
+  
+  
+  private func replace(oldLocation: ChapterLocation, with newLocation: ChapterLocation) {
+    guard
+      let oldLocation = oldLocation.toTPPBookLocation(),
+      let newLocation = newLocation.toTPPBookLocation() else { return }
+    registry.replaceGenericBookmark(oldLocation, with: newLocation, forIdentifier: book.identifier)
+  }
+}
+
+extension Array where Element == ChapterLocation {
```

Comment:
Maybe nitpicking, can we move this array extension and the ChapterLocation extension below to a separate file if they shouldn't be in NYPLAudiobookToolkit? Also would be great to add another ChapterLocation extension from TPPBookRegistry+Extensions.swift to that file as well )

Reply:
I agree with moving the ChapterLocationExtension to its own file. But I think that this should stay here because it is a very specific implementation that should only be used in the context that this class uses it.

Reply:
@vladimirfedorov  Updates here: https://github.com/ThePalaceProject/ios-audiobooktoolkit/pull/44